# Retail Data Analytics

## Initial Exploratory Data Analysis

Goal of this project is to predict the sales of a Retail outlet based on the historical data provided for sales. A retailer wants to know the sales of his outlet in different locations across the country. This gives him a best understanding of scaling his business and to use the best practices which improves the sales of outlets. We have been given two kinds of data, one of them have historical data of sales across different outlets. For the other one we need to predict the sales based on the different parameters given. As we move through the various steps, we build a model to predict the sales of Retail Store.

In the first step we are going to acquire data into dataframes for further processing.

In [30]:
import numpy as np
import pandas as pd
data_Train = pd.read_csv("Train_Retail.csv",index_col="Item_Identifier")
data_Test = pd.read_csv("Test_Retail.csv",index_col="Item_Identifier")

In [31]:
data_Train.shape

(8523, 11)

In [32]:
data_Test.shape

(5681, 10)

In last few steps we loaded the data from Test and Train datasets into the data frames data_Test and data_Train respectively. Shape function gives us the shape i.e. number of rows and columns present in the dataframe. Train dataset has 8523 rows and 11 columns and data_Test has 5681 rows and 10 columns in it.

While predicting an outcome it is always important to understand our dataset properly and know the effecting features in the dataset. Let's have a look into the dataset given for training and test data. 

In [33]:
data_Train.head

<bound method NDFrame.head of                  Item_Weight Item_Fat_Content  Item_Visibility  \
Item_Identifier                                                  
FDA15                  9.300          Low Fat         0.016047   
DRC01                  5.920          Regular         0.019278   
FDN15                 17.500          Low Fat         0.016760   
FDX07                 19.200          Regular         0.000000   
NCD19                  8.930          Low Fat         0.000000   
FDP36                 10.395          Regular         0.000000   
FDO10                 13.650          Regular         0.012741   
FDP10                    NaN          Low Fat         0.127470   
FDH17                 16.200          Regular         0.016687   
FDU28                 19.200          Regular         0.094450   
FDY07                 11.800          Low Fat         0.000000   
FDA03                 18.500          Regular         0.045464   
FDX32                 15.100          Regular 

In [34]:
data_Test.head

<bound method NDFrame.head of                  Item_Weight Item_Fat_Content  Item_Visibility  \
Item_Identifier                                                  
FDW58                 20.750          Low Fat         0.007565   
FDW14                  8.300              reg         0.038428   
NCN55                 14.600          Low Fat         0.099575   
FDQ58                  7.315          Low Fat         0.015388   
FDY38                    NaN          Regular         0.118599   
FDH56                  9.800          Regular         0.063817   
FDL48                 19.350          Regular         0.082602   
FDC48                    NaN          Low Fat         0.015782   
FDN33                  6.305          Regular         0.123365   
FDA36                  5.985          Low Fat         0.005698   
FDT44                 16.600          Low Fat         0.103569   
FDQ56                  6.590          Low Fat         0.105811   
NCC54                    NaN          Low Fat 

Head function gives us glimpse of data present in the Train and Test datasets. Although we can look at the data it is not so clear what are the different columns present in the dataset. By taking only one row of the dataset, lets examine the different columns. 

In [35]:
data_Train[:1]

Item_Weight Item_Fat_Content  Item_Visibility Item_Type  \
Item_Identifier                                                            
FDA15                    9.3          Low Fat         0.016047     Dairy   

                 Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
Item_Identifier                                                          
FDA15            249.8092            OUT049                       1999   

                Outlet_Size Outlet_Location_Type        Outlet_Type  \
Item_Identifier                                                       
FDA15                Medium               Tier 1  Supermarket Type1   

                 Item_Outlet_Sales  
Item_Identifier                     
FDA15                     3735.138

With this it is clear that we have the columns named "Item_Identifier","Item_Weight","Item_Fat_Content","Item_Visibility","Item_Type","Item_MRP","Outlet_Identifier",
"Outlet_Establishment_Year","Outlet_Size","Outlet_Location_Type","Outlet_Type","Item_Outlet_Sales" in our training dataset. Of all these attributes all the columns in the starting are Predictors for the outcome "Item_Outlet_Sales".

Lets do the same for Test Dataset.

In [36]:
data_Test[:1]

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
Item_Identifier                                                               
FDW58                  20.75          Low Fat         0.007565  Snack Foods   

                 Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
Item_Identifier                                                          
FDW58            107.8622            OUT049                       1999   

                Outlet_Size Outlet_Location_Type        Outlet_Type  
Item_Identifier                                                      
FDW58                Medium               Tier 1  Supermarket Type1

If we observe the attributes in the two datasets, we can observe that both test and train datasets have same type of columns except that in the test dataset Item_Outlet_Sales are missing, which we need to predict from the model we build from the train dataset. Now check for the datatypes of different columns in the dataset which will be helpful in further data processing operations. 

In [37]:
data_Train.dtypes

Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [38]:
data_Test.dtypes

Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
dtype: object

We can see that our dataset has statistical values in them and also we have few of the categorical columns. Lets check for any of the missing data items in different attributes for our datasets.  

In [39]:
data_Train.apply(lambda x: sum(x.isnull()))

Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In our training dataset we have the missing data values for the columns "Item_Weight" and "Outlet_Size" each of them having missing values for 1463 and 2410 number of rows. By checking the same for Test dataset. 

In [40]:
data_Test.apply(lambda x: sum(x.isnull()))

Item_Weight                   976
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  1606
Outlet_Location_Type            0
Outlet_Type                     0
dtype: int64

Our Test dataset also have missing values for "Item_Weight" and "Outlet_Size" columns. We are sure that our datasets have missing values in them. We need to deal with these missing values, leaving behind missing values have implications in building a successful model. Hence, it is often adviced to take the missing values into account and impute them properly. We shall deal with the missing values in our next section "Imputing the missing values". We have seen the attributes, datatypes and number of missing values. For better understanding the values in the dataset lets have a look into the statistics of our datasets.

In [41]:
data_Train.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

#### Some Important Observations from the Statistics of Training dataset:

1. Column "Item_Weight" has values ranging from 4.56 to 21.35 with a mean value of 12.86 (adjusted to 2 decimal points). Count is showing as 7060 although we have 8523 number of rows in our training dataset. This phenomenon is justified by missing value count of 1463.
2. "Item_Visibility" has values ranging from 0 to 0.33 with a mean value of 0.066. But, min value of 0 is not practically feasible in this column. Since an Item can't be sold unless it is displayed in the shelves. We will change these values in our next sections.
3. "Item_MRP" has values ranging from 31.29 to 266.89 with a mean value of 141. 
4. "Outlet_Establishment_Year" has values ranging from 1985 to 2009. From this we can see that, first store started is in 1985 and the most recent store started in 2009.

Doing the same for our Test dataset. 

In [42]:
data_Test.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year
count  4705.000000      5681.000000  5681.000000                5681.000000
mean     12.695633         0.065684   141.023273                1997.828903
std       4.664849         0.051252    61.809091                   8.372256
min       4.555000         0.000000    31.990000                1985.000000
25%       8.645000         0.027047    94.412000                1987.000000
50%      12.500000         0.054154   141.415400                1999.000000
75%      16.700000         0.093463   186.026600                2004.000000
max      21.350000         0.323637   266.588400                2009.000000

In our Test dataset we have total of 5681 number of rows. Of these we have missing values for the attribute named Item_Weight. In our next steps we will deal with the missing values in the dataset. 

## Dealing with Missing values in the dataset

Having missing values in the dataset complicates the process of predicting the output and create a model that well performs. In order to deal with the missing values on the first hand we need to identify cause for missing item values in the dataset. Most of the times these data points are not recorded at the source or ignored while collecting the data. If the data points are collected and ignored while taking it into consideration it's often a good idea to re-enter those values from the source. If they were not collected at the source we need to find a way to impute those missing values.

 There are many ways to deal with the missing values based on the methodology we use. One easy way is to impute those missing values with the mean values of that attribute type. Although practically it is a very simple way to deal with the missing values, the result of this method are often not so accurate. For the current project I'm going to first impute the missing values with their mean value and design a model. Later we will design model itself to specifically identify the missing values. 

In order to avoid the duplication of process to treat the missing values separately in the Test and Train datasets, we are going to merge those two datasets and do the necessary operations.

In [43]:
data = pd.concat([data_Train,data_Test])

In [44]:
data.shape

(14204, 11)

In the combined dataset we have around 14,204 number of rows. As seen in the previous sections, we have missing values for "Item_Weight". Let's impute these item weights by mean of the Item_Weights. 

In [49]:
data.mean()


Item_MRP                      141.004977
Item_Outlet_Sales            2181.288914
Item_Visibility                 0.065953
Item_Weight                    12.792854
Outlet_Establishment_Year    1997.830681
dtype: float64

In [51]:
data.fillna(data.mean())

Item_Fat_Content  Item_MRP  Item_Outlet_Sales  \
Item_Identifier                                                 
FDA15                    Low Fat  249.8092        3735.138000   
DRC01                    Regular   48.2692         443.422800   
FDN15                    Low Fat  141.6180        2097.270000   
FDX07                    Regular  182.0950         732.380000   
NCD19                    Low Fat   53.8614         994.705200   
FDP36                    Regular   51.4008         556.608800   
FDO10                    Regular   57.6588         343.552800   
FDP10                    Low Fat  107.7622        4022.763600   
FDH17                    Regular   96.9726        1076.598600   
FDU28                    Regular  187.8214        4710.535000   
FDY07                    Low Fat   45.5402        1516.026600   
FDA03                    Regular  144.1102        2187.153000   
FDX32                    Regular  145.4786        1589.264600   
FDS46                    Regular  119.6782        2145.207600   
FDF32                    Low Fat  196.4426        1977.426000   
FDP49                    Regular   56.3614        1547.319200   
NCB42                    Low Fat  115.3492        1621.888800   
FDP49                    Regular   54.3614         718.398200   
DRI11                    Low Fat  113.2834        2303.668000   
FDU02                    Low Fat  230.5352        2748.422400   
FDN22                    Regular  250.8724        3775.086000   
FDW12                    Regular  144.5444        4064.043200   
NCB30                    Low Fat  196.5084        1587.267200   
FDC37                    Low Fat  107.6938         214.387600   
FDR28                    Regular  165.0210        4078.025000   
NCD06                    Low Fat   45.9060         838.908000   
FDV10                    Regular   42.3112        1065.280000   
DRJ59                    low fat   39.1164         308.931200   
FDE51                    Regular   45.5086         178.434400   
FDC14                    Regular   43.6454         125.836200   
...                          ...       ...                ...   
FDD23                    Regular  186.1898        2181.288914   
FDP32                    Low Fat  126.7678        2181.288914   
FDO31                    Regular   80.2960        2181.288914   
FDQ57                    Low Fat  144.5760        2181.288914   
FDX32                    Regular  143.6786        2181.288914   
FDC39                    Low Fat  207.8296        2181.288914   
FDU55                    Low Fat  261.3278        2181.288914   
FDL45                    Low Fat  123.4704        2181.288914   
DRK37                    Low Fat  188.8530        2181.288914   
FDK22                    Low Fat  215.3850        2181.288914   
DRG37                    Low Fat  155.7972        2181.288914   
FDK22                    Low Fat  214.3850        2181.288914   
DRH36                    Low Fat   72.8696        2181.288914   
DRC36                    Regular  173.4054        2181.288914   
DRE03                    Low Fat   46.0718        2181.288914   
FDF34                    Regular  196.9084        2181.288914   
FDZ22                    Low Fat   82.1250        2181.288914   
FDC44                    Low Fat  115.1518        2181.288914   
FDN31                    Low Fat  188.0530        2181.288914   
FDO03                    Regular  229.4352        2181.288914   
FDA01                        reg   59.5904        2181.288914   
NCH42                    Low Fat  231.1010        2181.288914   
FDF46                    Low Fat  116.0834        2181.288914   
DRL35                    Low Fat   43.2770        2181.288914   
FDW46                    Regular   63.4484        2181.288914   
FDB58                    Regular  141.3154        2181.288914   
FDD47                    Regular  169.1448        2181.288914   
NCO17                    Low Fat  118.7440        2181.288914   
FDJ26                    Regular  214.6218        2